### PR TITLE
[Merged by Bors] - chore: update mk_all script and use it in CI

### PR DESCRIPTION
**Pull request by adomani**
Prompted by [this Zulip message](https://leanprover.zulipchat.com/#narrow/stream/116290-rss/topic/Recent.20Commits.20to.20mathlib4.3Amaster/near/430753770), I unified all the uses that I could find of "list all files in some dir and create an 'import file' for them".

In the process, I added a slight flexibility to the `mk_all` script: if you pass it a space-separated list of target dirs, it will produce 'import files' for each dir in the list.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: .github/workflows/build.yml.in [hunk 1]
```diff
@@ -61,25 +61,10 @@ jobs:
...
-
-      - name: update Counterexamples.lean
-        run: |
-          git ls-files 'Counterexamples/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Counterexamples.lean
-
-      - name: update Archive.lean
+      - name: import files in {Mathlib MathlibExtras Mathlib/Tactic Counterexamples Archive}.lean
         run: |
-          git ls-files 'Archive/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Archive.lean
+          chmod u+x scripts/mk_all.sh
```
Comment (Ruben-VandeVelde) [inline .github/workflows/build.yml.in:66]: Is this really necessary?

File: scripts/mk_all.sh [hunk 2]
```diff
@@ -1,6 +1,33 @@
...
+##
+##  are equivalent.
+
 cd "$(git rev-parse --show-toplevel)" || exit 1
-for gp in Mathlib MathlibExtras Mathlib/Tactic Counterexamples Archive
+
+# assign `targets` to be the input if provided, default to "the usual Mathlib suspects" otherwise
+if [ -n "${*}" ]
+then
+  targets=${*}
```
Comment (Ruben-VandeVelde) [inline scripts/mk_all.sh:24]: I don't know enough bash to review this, but would approve the CI changes.

Comment (adomani) [reply scripts/mk_all.sh:24]: If it helps, `-n` checks whether the variable is assigned and non empty. `$*` means "all arguments passed to the function".

File: .github/workflows/build.yml.in (hunk 1 above)
Comment (adomani) [reply .github/workflows/build.yml.in:66]: Not sure: I'll test once I am at a computer.

Comment (adomani) [reply .github/workflows/build.yml.in]: A few weeks later... you were right, it was not needed and I removed it!

File: .github/workflows/bors.yml [hunk 3]
```diff
@@ -75,21 +75,9 @@ jobs:
...
-      - name: update Mathlib/Tactic.lean
-        run: |
-          git ls-files 'Mathlib/Tactic/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Mathlib/Tactic.lean
-
-      - name: update Counterexamples.lean
-        run: |
-          git ls-files 'Counterexamples/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Counterexamples.lean
-
-      - name: update Archive.lean
+      - name: import files in {Mathlib Mathlib/Tactic Counterexamples Archive}.lean
```
Comment (grunweg) [inline .github/workflows/bors.yml:78]: How about one of these?
```suggestion
      - name: update import-only files {Mathlib Mathlib/Tactic Counterexamples Archive}.lean
 ```
```suggestion
      - name: update lists of files in {Mathlib Mathlib/Tactic Counterexamples Archive}.lean

File: .github/workflows/bors.yml [hunk 4]
```diff
@@ -216,7 +204,6 @@ jobs:
       - name: find `#`-commands
         id: hash_commands
         run: |
-          chmod u+x scripts/lint_hash_commands.sh
```
Comment (grunweg) [inline .github/workflows/bors.yml:219]: Let's revert this stray change (and simply land the rewrite of the hashCommandLinter instead).
```
         chmod u+x scripts/lint_hash_commands.sh

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -82,21 +82,9 @@ jobs:
...
-      - name: update Mathlib/Tactic.lean
-        run: |
-          git ls-files 'Mathlib/Tactic/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Mathlib/Tactic.lean
-
-      - name: update Counterexamples.lean
-        run: |
-          git ls-files 'Counterexamples/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Counterexamples.lean
-
-      - name: update Archive.lean
+      - name: import files in {Mathlib Mathlib/Tactic Counterexamples Archive}.lean
```
Comment (grunweg) [inline .github/workflows/build.yml:85]: As above, I suggest tweaking this description.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -223,7 +211,6 @@ jobs:
       - name: find `#`-commands
         id: hash_commands
         run: |
-          chmod u+x scripts/lint_hash_commands.sh
```
Comment (grunweg) [inline .github/workflows/build.yml:226]: Also, please revert this change.

File: .github/workflows/build.yml.in [hunk 7]
```diff
@@ -61,21 +61,9 @@ jobs:
...
-      - name: update Mathlib/Tactic.lean
-        run: |
-          git ls-files 'Mathlib/Tactic/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Mathlib/Tactic.lean
-
-      - name: update Counterexamples.lean
-        run: |
-          git ls-files 'Counterexamples/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Counterexamples.lean
-
-      - name: update Archive.lean
+      - name: import files in {Mathlib Mathlib/Tactic Counterexamples Archive}.lean
```
Comment (grunweg) [inline .github/workflows/build.yml.in:64]: Same here.

File: .github/workflows/build.yml.in [hunk 8]
```diff
@@ -202,7 +190,6 @@ jobs:
       - name: find `#`-commands
         id: hash_commands
         run: |
-          chmod u+x scripts/lint_hash_commands.sh
```
Comment (grunweg) [inline .github/workflows/build.yml.in:205]: And here.

File: .github/workflows/build_fork.yml [hunk 9]
```diff
@@ -79,21 +79,9 @@ jobs:
...
-      - name: update Mathlib/Tactic.lean
-        run: |
-          git ls-files 'Mathlib/Tactic/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Mathlib/Tactic.lean
-
-      - name: update Counterexamples.lean
-        run: |
-          git ls-files 'Counterexamples/*.lean' | LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Counterexamples.lean
-
-      - name: update Archive.lean
+      - name: import files in {Mathlib Mathlib/Tactic Counterexamples Archive}.lean
```
Comment (grunweg) [inline .github/workflows/build_fork.yml:82]: As above.

File: .github/workflows/build_fork.yml [hunk 10]
```diff
@@ -220,7 +208,6 @@ jobs:
       - name: find `#`-commands
         id: hash_commands
         run: |
-          chmod u+x scripts/lint_hash_commands.sh
```
Comment (grunweg) [inline .github/workflows/build_fork.yml:223]: As above.

File: .github/workflows/label_new_contributor.yml [hunk 11]
```diff
@@ -15,7 +15,7 @@ jobs:
       contents: read
     steps:
       - name: Label PR and report count
-        uses: actions/github-script@v5
+        uses: actions/github-script@v7
```
Comment (grunweg) [inline .github/workflows/label_new_contributor.yml:18]: If you rebase this PR/merge master, this change will go away :-)

File: .github/workflows/lint_and_suggest_pr.yml [hunk 12]
```diff
@@ -57,21 +57,9 @@ jobs:
 
       - uses: actions/checkout@v4
 
-      - name: update Mathlib.lean
+      - name: import files in {Mathlib Mathlib/Tactic Counterexamples Archive}.lean
```
Comment (grunweg) [inline .github/workflows/lint_and_suggest_pr.yml:60]: As above.

File: .github/workflows/nightly_detect_failure.yml [hunk 13]
```diff
@@ -114,7 +114,7 @@ jobs:
 
     - name: Check for matching bump/nightly-YYYY-MM-DD branch
       id: check_branch
-      uses: actions/github-script@v5
+      uses: actions/github-script@v7
```
Comment (grunweg) [inline .github/workflows/nightly_detect_failure.yml:117]: These will also go away when rebasing/merging master.

File: scripts/mk_all.sh [hunk 14]
```diff
@@ -1,6 +1,33 @@
 #! /usr/bin/env bash
+
```
Comment (grunweg) [inline scripts/mk_all.sh:2]: I personally would start with a one-sentence description, something like: "create or update a file with all imports in a given directory" or list of directories. (Or shamelessly copying from #11874, something like "Generate a file importing all the files of a Lean folder. By default, it generates the files for the libraries `Mathlib`, `Mathlib/Tactic`, `Archive` and `Counterexamples`.")

File: scripts/mk_all.sh [hunk 15]
```diff
@@ -1,6 +1,33 @@
...
+##  Standard use:
+##  ./scripts/mk_all.sh
+##
+##  updates `Mathlib` `Mathlib/Tactic` `Counterexamples` `Archive`
```
Comment (grunweg) [inline scripts/mk_all.sh:6]:
```suggestion
##  updates `Mathlib.lean`, `Mathlib/Tactic.lean`, `Counterexamples.lean` and `Archive.lean`

File: scripts/mk_all.sh [hunk 16]
```diff
@@ -1,6 +1,33 @@
...
+##
+##  Pro use:
```
Comment (grunweg) [inline scripts/mk_all.sh:8]: "Advanced use" maybe?

File: scripts/mk_all.sh (hunk 2 above)
Comment (grunweg) [reply scripts/mk_all.sh:24]: I've looked at this enough to convince myself this should do what it says.
In any case, #11874 offers rewriting this in Lean - so I am not worried about long-term maintainability. Short-term, this seems fine and like a clear step forward.

**Review by grunweg — COMMENTED** [review]
I think this is a good change and a clear step forward. Thank you for refactoring mathlib's CI!

Re-writing the `mk_all` script in Lean is another step forward, but I don't see a reason to block this step on that. I have only minor comments - with these addressed (most by simply merging master), I'll approve.

**Comment by grunweg** [comment]
I do wonder about the duplication in mathlib's CI - did you really have to write the same thing *four times*? But this is beyond this PR...

File: .github/workflows/bors.yml (hunk 3 above)
Comment (adomani) [reply .github/workflows/bors.yml]: I went for the first option: `update import-only files`

**Review by adomani — COMMENTED** [review]
Michael, as usual, thank you so much for your comments!

I have accepted them all and merged master: the PR is much cleaner now!

**Comment by grunweg** [comment]
Thanks for so quickly addressing my comments. I think this is worth trying out; see [my comment above](https://github.com/leanprover-community/mathlib4/pull/11849#pullrequestreview-2074603523) for details.
maintainer merge

File: scripts/mk_all.sh (hunk 2 above)
Comment (Ruben-VandeVelde) [reply scripts/mk_all.sh:28]: Thinking about this again, I don't see the use case for passing arguments to this script. Unless there's a strong reason for it that I haven't thought about, I'd revert that part. The main selling point of this script is that it's simple and fast, IMO.

Comment (adomani) [reply scripts/mk_all.sh:28]: The reason I added that option is for downstream projects that can target whichever folder they want.  This was before the lean script, though.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
